src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -148,41 +159,64 @@ private void CheckForShutdown()
         {
-            Debug.Assert(async);
-
             // Allocate an active request
             QuicStream? quicStream = null;
             Http3RequestStream? requestStream = null;
             ValueTask waitTask = default;
 
             try
             {
-                while (true)
+                try
                 {
-                    lock (SyncObj)
+                    while (true)
                     {
-                        if (_connection == null)
+                        lock (SyncObj)
                         {
-                            break;
+                            if (_connection == null)
+                            {
+                                break;
+                            }
+
+                            if (_connection.GetRemoteAvailableBidirectionalStreamCount() > 0)
+                            {
+                                quicStream = _connection.OpenBidirectionalStream();
+                                requestStream = new Http3RequestStream(request, this, quicStream);
+                                _activeRequests.Add(quicStream, requestStream);
+                                break;
+                            }
+
+                            waitTask = _connection.WaitForAvailableBidirectionalStreamsAsync(cancellationToken);
                         }
 
-                        if (_connection.GetRemoteAvailableBidirectionalStreamCount() > 0)
+                        if (HttpTelemetry.Log.IsEnabled() && !waitTask.IsCompleted && !queueDuration.IsActive)
                         {
-                            quicStream = _connection.OpenBidirectionalStream();
-                            requestStream = new Http3RequestStream(request, this, quicStream);
-                            _activeRequests.Add(quicStream, requestStream);
-                            break;
+                            // We avoid logging RequestLeftQueue if a stream was available immediately (synchronously)
+                            queueDuration = ValueStopwatch.StartNew();
                         }
-                        waitTask = _connection.WaitForAvailableBidirectionalStreamsAsync(cancellationToken);
-                    }
 
-                    // Wait for an available stream (based on QUIC MAX_STREAMS) if there isn't one available yet.
-                    await waitTask.ConfigureAwait(false);
+                        // Wait for an available stream (based on QUIC MAX_STREAMS) if there isn't one available yet.
+                        await waitTask.ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    if (HttpTelemetry.Log.IsEnabled() && queueDuration.IsActive)
+                    {
+                        HttpTelemetry.Log.Http30RequestLeftQueue(queueDuration.GetElapsedTime().TotalMilliseconds);
```

Comment:
In case the connection gets closed in the meantime, the request will get retried on a different connection (line 224). I'm not sure what counts into the "waiting time in the queue" in such case, but this will report as request left the queue.

Reply:
This should be rare-ish to happen, but yes, you may see multiple `RequestLeftQueue` events for a given request.

Similarly with 1.1, there is a race where we would retry if the server disconnects [here](https://github.com/dotnet/runtime/blob/dc9979136cf0ff1addbc1a4c0f06ea4b65f002d4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L587-L593), logging `RequestLeftQueue` twice.
Similarly with resets on HTTP/2, we may re-insert the request to the queue.

I'm not sure we can/should improve that somehow. I feel it's still valuable to include the time on queue here even if the request will just end up back in the queue.

Reply:
If we do similar thing with H/1.1 and H/2 then this is fine. I'm mostly trying to make sure we keep similar behavior between the protocols so that one counter doesn't mean something else based on HTTP version.